marlowe-playground-client/src/Marlowe/Blockly.purs

```diff
@@ -999,278 +1001,354 @@ toDefinition blockType@(ValueType UseValueValueType) =
-  traverse_ (\t -> mkGenFun t (blockDefinition t generator)) valueTypes
-  traverse_ (\t -> mkGenFun t (blockDefinition t generator)) actionTypes
-  traverse_ (\t -> mkGenFun t (boundsDefinition generator)) [ BoundsType ]
-  pure generator
-
-class HasBlockDefinition a b | a -> b where
-  blockDefinition :: a -> Generator -> Block -> Either String b
-
-baseContractDefinition :: Generator -> Block -> Either String (Term Contract)
-baseContractDefinition g block = case statementToCode g block (show BaseContractType) of
-  Either.Left _ -> pure $ Hole "contract" Proxy NoLocation
-  Either.Right s -> parse (mkDefaultTerm <$> Parser.contract) s
-
-getAllBlocks :: Block -> Array Block
-getAllBlocks currentBlock =
-  currentBlock
-    : ( case nextBlock currentBlock of
-          Nothing -> []
-          Just nextBlock' -> getAllBlocks nextBlock'
-      )
+explainParseTermError :: ParseTermError -> String
+explainParseTermError (InvalidBlock (BDom.Block block) expectedType) = "Block with id " <> show block.id <> " and type " <> show block.type <> " is not a valid " <> expectedType <> " type"
+
+explainParseTermError (MissingArgument (BDom.Block block) attr) = "Block " <> show block.id <> " is missing required argument " <> show attr
+
+explainParseTermError (InvalidChildType childType) = "Attribute should be of type " <> childType
 
-caseDefinition :: Generator -> Block -> Either String (Term Case)
-caseDefinition g block =
+explainParseTermError (InvalidFieldCast fieldValue castType) = "Could not convert value " <> show fieldValue <> " to " <> castType
+
+explainParseTermError (ErrorInChild (BDom.Block block) attr err) = "Error in " <> block.type <> "." <> attr <> explainNested err
+  where
+  explainNested (ErrorInChild (BDom.Block block') attr' err') = " -> " <> block'.type <> "." <> attr' <> explainNested err'
+
+  explainNested err' = ": " <> explainParseTermError err'
+
+getRequiredAttribute :: forall m. MonadThrow ParseTermError m => String -> BDom.Block -> m BDom.BlockChild
+getRequiredAttribute attr b@(BDom.Block block) = case (Object.lookup attr block.children) of
+  Nothing -> throwError $ MissingArgument b attr
+  Just child -> pure child
+
+asStatement :: forall m. MonadThrow ParseTermError m => BDom.BlockChild -> m (Array BDom.Block)
+asStatement child = case child of
+  BDom.Statement blocks -> pure blocks
+  _ -> throwError $ InvalidChildType "Statement"
+
+asSingleStatement :: forall m. MonadThrow ParseTermError m => BDom.BlockChild -> m BDom.Block
+asSingleStatement child = do
+  statements <- asStatement child
+  if length statements /= 1 then
+    throwError $ InvalidChildType "Statement with a single child"
+  else
+    pure $ unsafePartial $ UnsafeArray.head statements
+
+asValue :: forall m. MonadThrow ParseTermError m => BDom.BlockChild -> m BDom.Block
+asValue child = case child of
+  BDom.Value block -> pure block
+  _ -> throwError $ InvalidChildType "Value"
```

Comment:
personally I think top level case looks better as a pattern match in the left hand side of the function instead
`asValue (BDom.Value block) = pure block`

Reply:
I normally do it this way for small functions as purty adds an empty line between definitions... For stuff like handleAction I like that empty line, but for small helpers is "distracting"